TypeForm incremental inconsistencies

**Bug Report**

The following example does not error when running mypy after deleting `.mypy_cache` or when running with `--no-incremental` but does error after editing `b.py`.
It seems to be caching `TypeForm` as `type`.

**To Reproduce**

1) Download the code either from the zip archive or code blocks.
2) Run `mypy`. It outputs `Success: no issues found in 2 source files`
3) Make a whitespace change to `b.py` (eg add a new line at the end)
4) Run `mypy`. It outputs the following
```
src\b.py:3: error: Argument 1 to "test" has incompatible type "str"; expected "type[Never]"  [arg-type]
Found 1 error in 1 file (checked 2 source files)
```
5) Deleting `.mypy_cache` and running `mypy` or running `mypy --no-incremental` outputs the same as 2.


[mypy.zip](https://github.com/user-attachments/files/28592195/mypy.zip)

```python
# src/a.py
from typing import TypeVar
from typing_extensions import TypeForm

T = TypeVar('T')

def test(arg: TypeForm[T]) -> None:
    raise NotImplementedError
```

```py
# src/b.py
from a import test

test("int")
```

```ini

; mypy.ini
[mypy]
disallow_untyped_defs = True
check_untyped_defs = True
warn_return_any = True
python_version = 3.14
explicit_package_bases = True
mypy_path = $MYPY_CONFIG_FILE_DIR/src
files =
    src
enable_incomplete_feature = TypeForm
```

**Expected Behavior**

I expect the output from mypy to be consistent regardless of caching.

**Your Environment**

- Mypy version used: 2.1.0
- Python version used: 3.14
- typing-extensions: 4.15.0

--- a/mypy/types.py
+++ b/mypy/types.py
@@ -3688,11 +3688,12 @@ class TypeType(ProperType):
     def write(self, data: WriteBuffer) -> None:
         write_tag(data, TYPE_TYPE)
         self.item.write(data)
+        write_bool(data, self.is_type_form)
         write_tag(data, END_TAG)
 
     @classmethod
     def read(cls, data: ReadBuffer) -> Type:
-        ret = TypeType.make_normalized(read_type(data))
+        ret = TypeType.make_normalized(read_type(data), is_type_form=read_bool(data))
         assert read_tag(data) == END_TAG
         return ret
 

--- a/test-data/unit/check-incremental.test
+++ b/test-data/unit/check-incremental.test
@@ -8116,3 +8116,26 @@ tmp/m.py:5: note: Revealed type is "builtins.str"
 tmp/m.py:4: note: Revealed type is "def () -> builtins.int"
 tmp/m.py:5: error: Incompatible types in assignment (expression has type "type[A]", variable has type "int")
 tmp/m.py:6: note: Revealed type is "builtins.str"
+
+[case testTypeFormWorksInIncrementalMode]
+import b
+[file a.py]
+from typing import TypeVar
+from typing_extensions import TypeForm
+
+T = TypeVar("T")
+def test(arg: TypeForm[T]) -> T: ...
+
+[file b.py]
+from a import test
+test("int")
+
+[file b.py.2]
+from a import test
+reveal_type(test("int"))
+
+[builtins fixtures/primitives.pyi]
+[typing fixtures/typing-full.pyi]
+[out]
+[out2]
+tmp/b.py:2: note: Revealed type is "builtins.int"

